plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Outline.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Ide.Plugin.Cabal.Outline
+  ( moduleOutline,
+  )
+where
+
+import Control.Monad.IO.Class
+import Data.Maybe
+import Data.Text qualified as T
+import Data.Text.Encoding (decodeASCII, decodeLatin1)
+import Development.IDE.Core.Rules
+import Development.IDE.Core.Shake (IdeState (shakeExtras), runIdeAction, useWithStaleFast)
+import Development.IDE.Types.Location (toNormalizedFilePath')
+import Distribution.Fields.Field
+  ( Field (Field, Section),
+    FieldLine (FieldLine),
+    Name (Name),
+    SectionArg (SecArgName, SecArgOther, SecArgStr),
+  )
+import Distribution.Parsec.Position (Position)
+import Ide.Plugin.Cabal.Completion.Types (ParseCabalFields (..), cabalPositionToLSPPosition)
+import Ide.Plugin.Cabal.Orphans ()
+import Ide.Types
+import Language.LSP.Protocol.Message qualified as LSP
+import Language.LSP.Protocol.Types qualified as LSP
+
+moduleOutline :: PluginMethodHandler IdeState LSP.Method_TextDocumentDocumentSymbol
+moduleOutline ideState _ LSP.DocumentSymbolParams {_textDocument = LSP.TextDocumentIdentifier uri} =
+  case LSP.uriToFilePath uri of
+    Just (toNormalizedFilePath' -> fp) -> do
+      mFields <- liftIO $ runIdeAction "cabal-plugin.fields" (shakeExtras ideState) (useWithStaleFast ParseCabalFields fp)
+      case fmap fst mFields of
+        Just fieldPositions -> pure $ LSP.InR (LSP.InL allSymbols)
+          where
+            allSymbols = mapMaybe documentSymbolForField fieldPositions
+        Nothing -> pure $ LSP.InL []
+    Nothing -> pure $ LSP.InL []
+
+documentSymbolForField :: Field Position -> Maybe LSP.DocumentSymbol
+documentSymbolForField (Field (Name pos fieldName) fieldLines) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeASCII fieldName,
+        LSP._kind = LSP.SymbolKind_Object,
+        LSP._children = Just $ mapMaybe documentSymbolForFieldLine fieldLines
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII fieldName
+documentSymbolForField (Section (Name pos fieldName) sectionArgs fields) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeASCII fieldName,
+        LSP._kind = LSP.SymbolKind_Object,
+        LSP._children =
+          Just
+            ( mapMaybe documentSymbolForField fields
+                ++ mapMaybe documentSymbolForSectionArgs sectionArgs
+            )
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeASCII fieldName
+
+documentSymbolForSectionArgs :: SectionArg Position -> Maybe LSP.DocumentSymbol
+documentSymbolForSectionArgs (SecArgName pos identifier) =
+  Just
+    (defDocumentSymbol range)
+      { LSP._name = decodeLatin1 identifier,
```

Comment:
please document your choices for decoding in a Note

Reply:
Yes, that was motivated by various [unusual data](https://hackage.haskell.org/package/Cabal-syntax-3.10.2.0/docs/Distribution-Fields-Parser.html#g:1) that could be stored in the `SecArgName`'s _name.
I couldn't find an example of an edge case, though.

Reply:
I think this should be [`decodeUtf8`](https://hackage.haskell.org/package/text-2.1.1/docs/Data-Text-Encoding.html#v:decodeUtf8) either way?